File: python/Stack/min_stack.py
class MinStack:
    """
    Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
    
    Approach:
    1. Use two stacks:
       - main stack for normal operations
       - minStack to track minimum values
    2. For each push:
       - Push to main stack
       - Push min(current value, current min) to minStack
    3. For each pop:
       - Pop from both stacks
    4. For getMin:
       - Return top of minStack
    
    Time Complexity:
    - All operations are O(1)
    
    Space Complexity: O(n)
    - We store each element in both stacks
    """
    def __init__(self):
        self.stack = []  # Main stack
        self.minStack = []  # Stack to track minimum values

    def push(self, val: int) -> None:
        self.stack.append(val)
        # If minStack is empty, val is the minimum
        # Otherwise, compare with current minimum
        if not self.minStack:
            self.minStack.append(val)
        else:
            self.minStack.append(min(val, self.minStack[-1]))

    def pop(self) -> None:
        # Pop from both stacks
        self.stack.pop()
        self.minStack.pop()

    def top(self) -> int:
        # Return top element of main stack
        return self.stack[-1]

    def getMin(self) -> int:
        # Return top element of minStack (current minimum)
        return self.minStack[-1]

# -------------------
# Test Cases
# -------------------
def test_min_stack():
    # Test case 1: Basic push/pop/min
    minStack = MinStack()
    minStack.push(-2)
    minStack.push(0)
    minStack.push(-3)
    assert minStack.getMin() == -3
    minStack.pop()
    assert minStack.top() == 0
    assert minStack.getMin() == -2
    # Test case 2: Increasing order
    minStack = MinStack()
    minStack.push(1)
    minStack.push(2)
    minStack.push(3)
    assert minStack.getMin() == 1
    minStack.pop()
    assert minStack.getMin() == 1
    # Test case 3: Decreasing order
    minStack = MinStack()
    minStack.push(3)
    minStack.push(2)
    minStack.push(1)
    assert minStack.getMin() == 1
    minStack.pop()
    assert minStack.getMin() == 2
    # Test case 4: All same
    minStack = MinStack()
    minStack.push(5)
    minStack.push(5)
    minStack.push(5)
    assert minStack.getMin() == 5
    minStack.pop()
    minStack.pop()
    assert minStack.getMin() == 5
    print("All test cases passed!")

if __name__ == "__main__":
    test_min_stack()

File: C++/Stack/min_stack.cpp
#include <bits/stdc++.h>
using namespace std;

class MinStack {
private:
    /*
     * Use two stacks:
     * mainStack  - stores actual values
     * minStack   - tracks minimum at each level
     */
    stack<int> mainStack;  // Main stack
    stack<int> minStack;   // Minimum value stack

public:
    /*
     * Initialize empty MinStack
     * Time: O(1)
     */
    MinStack() {
        // Stacks are automatically initialized empty
    }
    
    /*
     * Push value onto stack and update minimum
     * Time: O(1)
     */
    void push(int val) {
        mainStack.push(val);
        
        // If minStack empty, val is minimum
        // Otherwise, push minimum of val and current minimum
        if (minStack.empty()) {
            minStack.push(val);
        } else {
            minStack.push(min(val, minStack.top()));
        }
    }
    
    /*
     * Remove top element from both stacks
     * Time: O(1)
     */
    void pop() {
        mainStack.pop();
        minStack.pop();
    }
    
    /*
     * Return top element of main stack
     * Time: O(1)
     */
    int top() {
        return mainStack.top();
    }
    
    /*
     * Return current minimum (top of minStack)
     * Time: O(1)
     */
    int getMin() {
        return minStack.top();
    }
};

/*
 * Detailed C++ Syntax Explanation for Beginners:
 * 
 * 1. Class Structure:
 *    private:           // Members only accessible within class
 *    public:            // Members accessible from outside
 * 
 * 2. Stack Declaration:
 *    stack<int>         // STL stack container for integers
 *    stack.push(val)    // Add element to top
 *    stack.pop()        // Remove top element
 *    stack.top()        // Access top element
 *    stack.empty()      // Check if stack is empty
 * 
 * 3. Constructor:
 *    MinStack()         // Called when object is created
 *    - No explicit initialization needed
 * 
 * 4. Standard Library:
 *    min(a, b)         // Returns smaller of two values
 * 
 * Note: This implementation:
 * - Uses O(n) extra space for minStack
 * - Provides O(1) time for all operations
 * - Maintains minimum value without searching
 */
